src/lib/shadow-shim-helper-rs/src/scmutex.rs

```diff
@@ -265,27 +299,44 @@ where
 
 #[cfg(test)]
 mod rkyv_tests {
-    use std::{ops::DerefMut, sync::Arc};
+    use std::pin::Pin;
+    use std::sync::Arc;
+
+    use rkyv::string::ArchivedString;
 
     use super::*;
 
+    /// Taking the AlignedVec here instead of a slice avoids a gray safety area
+    /// that violates miri's current Stacked Borrows check.
+    /// https://github.com/rkyv/rkyv/issues/303
+    ///
+    /// SAFETY:
+    /// * `vec` must contain a value of type `T` at the end.
+    unsafe fn archived_root<T>(vec: &rkyv::AlignedVec) -> &T::Archived
+    where
+        T: rkyv::Archive,
+    {
```

Comment:
I think the general idea of your suggestion would work, but afaict in addition to copying the data `AlignedVec::into_boxed_slice` doesn't guarantee that the data is correctly aligned. I think we'd need to create our own aligned memory allocation and copy out of the `AlignedVec` ourselves. I agree that `AlignedVec` could use an `into_raw` method that just handed over the internal allocation. 

Fortunately I *think* we don't have to bother with this since [AlignedVec::as_ptr](https://docs.rs/rkyv/0.5.0/rkyv/struct.AlignedVec.html#method.as_ptr) explicitly allows the pointer to be cast to a `T` with internal mutability via `UnsafeCell`; added some extra explanation here.